sycl-fusion/common/include/Kernel.h

```diff
@@ -147,7 +147,45 @@ struct SYCLArgumentDescriptor {
 /// List of SYCL/OpenCL kernel attributes.
 using AttributeList = std::vector<SYCLKernelAttribute>;
 
-using Indices = std::array<size_t, 3>;
+///
+/// Class to model a three-dimensional index.
+class Indices {
+public:
+  static constexpr size_t size() { return Size; }
+
+  constexpr Indices() : Values{0, 0, 0} {}
+  constexpr Indices(size_t V1, size_t V2, size_t V3) : Values{V1, V2, V3} {}
+
+  constexpr const size_t *begin() const { return Values; }
+  constexpr const size_t *end() const { return Values + Size; }
+  constexpr size_t *begin() { return Values; }
+  constexpr size_t *end() { return Values + Size; }
+
+  constexpr const size_t &operator[](int Idx) const { return Values[Idx]; }
+  constexpr size_t &operator[](int Idx) { return Values[Idx]; }
+
+  friend bool operator==(const Indices &A, const Indices &B) {
+    return std::equal(A.begin(), A.end(), B.begin());
+  }
+
+  friend bool operator!=(const Indices &A, const Indices &B) {
+    return !(A == B);
+  }
+
+  friend bool operator<(const Indices &A, const Indices &B) {
+    return std::lexicographical_compare(A.begin(), A.end(), B.begin(), B.end(),
+                                        std::less<size_t>{});
+  }
+
+  friend bool operator>(const Indices &A, const Indices &B) {
+    return std::lexicographical_compare(A.begin(), A.end(), B.begin(), B.end(),
+                                        std::greater<size_t>{});
+  }
+
+private:
+  static constexpr auto Size = 3;
```

Comment:
Is the type of `Size` actually `size_t`, if you use `auto` and `3` as literal initializer without any suffix (e.g., `ul`)?

Reply:
Right, no, it's `int`. Fixed.